airbyte/secrets/google_gsm.py

```diff
@@ -0,0 +1,209 @@
+# Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+"""Secret manager that retrieves secrets from Google Secrets Manager (GSM).
+
+Usage Example:
+
+```python
+gsm_secrets_manager = GoogleGSMSecretManager(
+    project=AIRBYTE_INTERNAL_GCP_PROJECT,
+    credentials_json=ab.get_secret("GCP_GSM_CREDENTIALS"),
+)
+first_secret: SecretHandle = next(
+    gsm_secrets_manager.fetch_connector_secrets(
+        connector_name=connector_name,
+    ),
+    None,
+)
+
+print(f"Found '{connector_name}' credential secret '${first_secret.secret_name}'.")
+return first_secret.get_value().parse_json()
+```
+
+More compact example:
+
+```python
+gsm_secrets_manager = GoogleGSMSecretManager(
+    project=AIRBYTE_INTERNAL_GCP_PROJECT,
+    credentials_json=ab.get_secret("GCP_GSM_CREDENTIALS"),
+)
+connector_config: dict = (
+    next(
+        gsm_secrets_manager.fetch_connector_secrets(
+            connector_name=connector_name,
+        ),
+        None,
+    )
+    .get_value()
+    .parse_json()
+)
+```
+"""
+
+from __future__ import annotations
+
+import json
+import os
+from pathlib import Path
+from typing import TYPE_CHECKING
+
+from google.cloud import secretmanager_v1 as secretmanager
+
+from airbyte import exceptions as exc
+from airbyte.secrets.base import SecretHandle, SecretSourceEnum, SecretString
+from airbyte.secrets.custom import CustomSecretManager
+
+
+if TYPE_CHECKING:
+    from collections.abc import Iterable
+
+    from google.cloud.secretmanager_v1.services.secret_manager_service.pagers import (
+        ListSecretsPager,
+    )
+
+
+class GoogleGSMSecretManager(CustomSecretManager):
+    """Secret manager that retrieves secrets from Google Secrets Manager (GSM).
+
+    This secret manager is not enabled by default. To use it, you must provide the project ID and
+    the credentials for a service account with the necessary permissions to access the secrets.
+    """
```

Comment:
Todo: make a note about how this implementation has additional capabilities beyond the base class.

Reply:
```suggestion
class GoogleGSMSecretManager(CustomSecretManager):
    """Secret manager that retrieves secrets from Google Secrets Manager (GSM).

    This class inherits from `CustomSecretManager` and also adds methods 
    that are specific to this implementation: `fetch_secrets()`,
    `fetch_secrets_by_label()` and `fetch_connector_secrets()`.

    This secret manager is not enabled by default. To use it, you must provide the project ID and
    the credentials for a service account with the necessary permissions to access the secrets.
    
    The `fetch_connector_secret()` method assumes a label name of `connector`
    matches the name of the connector (`source-github`, `destination-snowflake`, etc.)
    """
```